docs/design/asset-canister-interface.md

```diff
@@ -0,0 +1,508 @@
+
+### Content Encoding Selection
+
+When retrieving an asset, the caller specifies a list of acceptable [content encodings](#content_encoding). The asset canister will select the first suitable[^1] content encoding from this list.
+
+### Large Assets
+
+While the size of any given asset content encoding is limited only by the canister's available memory, the amount of data that can be passed or returned in a single method call is limited. For this reason, the interface provides for data upload and retrieval in smaller pieces, called "chunks".
+
+The size of each chunk is limited by the message ingress limit.
+
+## Storing Assets
+
+### Batch Updates
+
+The usual method of updating data in the asset canister is by calling the following methods:
+1. [create_batch()](#method-create_batch) once.
+2. [create_chunk()](#method-create_chunk) one or more times, which can occur concurrently.
+3. [commit_batch()](#method-commit_batch) zero or more times with `batch_id: 0`.
+4. [commit_batch()](#method-commit_batch) once with the batch ID from step 1, which indicates the batch is complete.
+
+The reason for multiple rather than single calls to [commit_batch][#method-commit_batch] is that certificate computation for an entire batch may exceed per-message computation limits.
+
+### Batch Updates By Proposal
+
+If an SNS controls an asset canister, it can update the assets by proposal. In this scenario, there are two principals:
+- The Preparer, which must have the [Prepare](#permission-prepare) permission. This principal prepares the proposal by uploading data and proposing changes to be committed.
+- The Committer, which must have the [Commit](#permission-commit) permission. This principal commits the previously-proposed changes.
+
+In this scenario, the Preparer calls the following methods:
+1. [create_batch()](#method-create_batch) once.
+2. [create_chunk()](#method-create_chunk) one or more times, which can occur concurrently.
+3. [propose_commit_batch()](#method-propose_commit_batch) once.
+4. [compute_evidence()](#method-compute_evidence) until the method returns `Some(evidence)`.
+
+The Preparer then furnishes the Committer with the following information:
+- the batch ID
+- the computed evidence
+
+The Committer then calls the following method upon approval of the proposal:
+1. [commit_proposed_batch()](#method-commit_proposed_batch)
+
+If the proposal is not approved, the Preparer must call [delete_batch()](#method-delete_batch).
+
+### Individual Updates
+
+It is also possible to upload a single content encoding of a single asset by calling the [store()](#method-store) method.  The size of the content encoding must not exceed the message ingress limit.
+
+## Type Reference
+
+### Asset
+
+#### Key
+
+The `key` is a case-sensitive string that identifies the asset. By convention, all asset keys begin with a forward slash. For example, `/index.html`.
```

Comment:
my previous comment alluded to this definition. perhaps a link to this section from up above would suffice.

Reply:
But I do still think a little more prose above will help, even if it duplicates some of what is here.